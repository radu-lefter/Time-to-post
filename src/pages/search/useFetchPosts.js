import { useEffect, useState } from 'react';

const NUM_POSTS_TO_FETCH = 500;
const MAX_NUM_POSTS_PER_PAGE = 100;

/**
 * The reddit endpoint that we fetch the top posts from uses pagination. We can fetch a maximum
 * number of 100 posts per page. In order to fetch the first 500 posts we use this recursive
 * function. Until the last page or the required number of posts has been reached we continue
 * to fetch more posts.
 *
 * @param {string} subreddit the name of the subreddit
 * @param {array} previousPosts the posts that have already been loaded
 *    (only to be used in recursive calls)
 * @param {string} after the id of the last post used for pagination
 *    (only to be used in recursive calls)
 * @returns {array} list of top 500 posts for subreddit
 */
export async function fetchPaginatedPosts(
  subreddit,
  abortController,
  previousPosts = [],
  after = null,
) {
  let url = `https://www.reddit.com/r/${subreddit}/top.json?t=year&limit=${MAX_NUM_POSTS_PER_PAGE}`;
  if (after) {
    url += `&after=${after}`;
  }
  const response = await fetch(url, { signal: abortController.signal });
  const { data } = await response.json();
  const allPosts = previousPosts.concat(data.children);

  const noMorePosts = data && data.dist < MAX_NUM_POSTS_PER_PAGE;
  const limitReached = allPosts.length >= NUM_POSTS_TO_FETCH;
  if (noMorePosts || limitReached) {
    return allPosts;
  }

  return fetchPaginatedPosts(subreddit, abortController, allPosts, data.after);
}

/**
 * Builds an object containing posts per day of week and hour to create the heatmap.
 * Each entry obj[dayOfWeek][hour] contains an array of posts.
 * dayOfWeek is a number between 0 and 6, hour a number between 0 and 23.
 *
 * @param {array} posts the concatenated list of posts returned from fetchPaginatedPosts
 * @returns {array} nested 3D array that contains the posts grouped by week day and hour
 */
function groupPostsPerDayAndHour(posts) {
  const postsPerDay = Array(7)
    .fill()
    .map(() => Array(24).fill().map(() => []));

  posts.forEach((post) => {
    const createdAtDate = new Date(post.data.created_utc * 1000);
    const dayOfWeek = createdAtDate.getDay();
    const hour = createdAtDate.getHours();

    postsPerDay[dayOfWeek][hour].push({
      createdAt: createdAtDate,
      title: post.data.title,
      url: `https://reddit.com${post.data.permalink}`,
      score: post.data.score,
      numComments: post.data.num_comments,
      author: post.data.author,
      authorId: post.data.author_fullname,
    });
  });

  return postsPerDay;
}

function useFetchPosts(subreddit) {
  const [postsPerDay, setPostsPerDay] = useState([]);
  const [status, setStatus] = useState('pending');

  useEffect(() => {
    const abortController = new AbortController();
    setStatus('pending');

    fetchPaginatedPosts(subreddit, abortController)
      .then((posts) => groupPostsPerDayAndHour(posts))
      .then((newpostsPerDay) => {
        setPostsPerDay(newpostsPerDay);
        setStatus('resolved');
      })
      .catch(() => {
        if (!abortController.signal.aborted) {
          setStatus('rejected');
        }
      });

    return () => abortController.abort();
  }, [subreddit]);

  return {
    isLoading: status === 'pending',
    hasError: status === 'rejected',
    postsPerDay,
  };
}

export default useFetchPosts;