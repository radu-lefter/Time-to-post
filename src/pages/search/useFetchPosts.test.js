import { renderHook } from '@testing-library/react-hooks';
import useFetchPosts from './useFetchPosts';

const getNumPosts = (nestedPostsArray) => nestedPostsArray.reduce(
  (numTotal, postsPerDay) => postsPerDay.reduce(
    (numPerDay, postsPerHour) => numPerDay + postsPerHour, numTotal,
  ),
  0,
);

test('loads 500 top posts from the Reddit API', async () => {
  const { result, waitForNextUpdate } = renderHook(() => useFetchPosts('500-posts'));

  expect(result.current.isLoading).toBe(true);
  expect(result.current.postsPerDay).toEqual([]);

  await waitForNextUpdate({ timeout: 10000 });

  expect(result.current.isLoading).toBe(false);
  expect(getNumPosts(result.current.postsPerDay)).toEqual(500);
  expect(result.current.postsPerDay).toMatchSnapshot();
});

test('stops loading when less than 500 posts are available', async () => {
  const { result, waitForNextUpdate } = renderHook(() => useFetchPosts('less-than-500-posts'));

  await waitForNextUpdate();

  expect(result.current.isLoading).toBe(false);
  expect(getNumPosts(result.current.postsPerDay)).toEqual(270);
});

test('returns error when a request fails', async () => {
  const { result, waitForNextUpdate } = renderHook(() => useFetchPosts('failing-request'));

  await waitForNextUpdate();

  expect(result.current.isLoading).toBe(false);
  expect(result.current.hasError).toEqual(true);
});