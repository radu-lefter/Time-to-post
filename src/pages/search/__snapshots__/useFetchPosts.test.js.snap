// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loads 500 top posts from the Reddit API 1`] = `
Array [
  Array [
    0,
    3,
    0,
    3,
    0,
    2,
    2,
    0,
    1,
    4,
    0,
    3,
    2,
    3,
    7,
    9,
    2,
    7,
    1,
    2,
    1,
    1,
    1,
    3,
  ],
  Array [
    2,
    1,
    0,
    1,
    3,
    1,
    5,
    3,
    4,
    4,
    1,
    2,
    5,
    0,
    6,
    12,
    6,
    6,
    4,
    2,
    5,
    3,
    3,
    2,
  ],
  Array [
    1,
    3,
    1,
    1,
    0,
    2,
    1,
    1,
    4,
    0,
    1,
    3,
    6,
    9,
    6,
    7,
    5,
    7,
    2,
    4,
    4,
    4,
    3,
    0,
  ],
  Array [
    1,
    2,
    2,
    0,
    3,
    1,
    3,
    1,
    3,
    4,
    2,
    4,
    5,
    5,
    7,
    5,
    4,
    5,
    1,
    4,
    3,
    3,
    3,
    3,
  ],
  Array [
    0,
    2,
    0,
    2,
    0,
    1,
    0,
    3,
    1,
    3,
    3,
    5,
    5,
    7,
    5,
    5,
    7,
    4,
    3,
    2,
    1,
    7,
    5,
    3,
  ],
  Array [
    2,
    2,
    3,
    3,
    0,
    2,
    1,
    2,
    0,
    1,
    3,
    4,
    4,
    4,
    4,
    6,
    4,
    4,
    1,
    4,
    7,
    3,
    2,
    2,
  ],
  Array [
    1,
    4,
    0,
    3,
    0,
    2,
    1,
    2,
    1,
    5,
    5,
    2,
    5,
    3,
    5,
    6,
    1,
    5,
    7,
    4,
    2,
    1,
    4,
    2,
  ],
]
`;
